File: apps/ingestion/src/producer/topics.ts
/**
 * Redpanda topic name constants.
 *
 * These are the canonical topic names established by ADR 0008
 * (Adopt Redpanda as Immutable Temporal Snapshot Ledger).
 *
 * Sprint 02 publishes only to SNAPSHOTS_RAW.
 * The remaining topics are defined here for completeness and to avoid
 * string literals being scattered across the codebase.
 */
export const TOPICS = {
  /** Raw decoded GTFS-RT snapshot payloads — primary target for Sprint 02 */
  SNAPSHOTS_RAW: 'transit.snapshots.raw',

  /**
   * Cleaned and structurally validated operational state snapshots.
   * Populated by downstream normalisation consumers in later sprints.
   */
  SNAPSHOTS_NORMALIZED: 'transit.snapshots.normalized',

  /**
   * Computed state transitions derived between consecutive snapshots.
   * Populated by downstream delta consumers in later sprints.
   */
  STATE_DELTAS: 'transit.state.deltas',

  /**
   * Derived observability and resilience metrics.
   * Populated by downstream metric consumers in later sprints.
   */
  METRICS_OPERATIONAL: 'transit.metrics.operational',
} as const;

export type TopicName = (typeof TOPICS)[keyof typeof TOPICS];

/**
 * Topic configuration for admin creation.
 *
 * Sprint 02 uses 1 partition per topic — ADR 0008 explicitly defers
 * partitioning specialisation until operational replay patterns are
 * validated empirically.
 *
 * Retention: 7 days for raw snapshots (168 hours × 3600 × 1000 ms).
 */
export const TOPIC_CONFIGS: Record<
  TopicName,
  { numPartitions: number; replicationFactor: number; retentionMs: number }
> = {
  [TOPICS.SNAPSHOTS_RAW]: {
    numPartitions: 1,
    replicationFactor: 1,
    retentionMs: 7 * 24 * 60 * 60 * 1000, // 7 days
  },
  [TOPICS.SNAPSHOTS_NORMALIZED]: {
    numPartitions: 1,
    replicationFactor: 1,
    retentionMs: 7 * 24 * 60 * 60 * 1000,
  },
  [TOPICS.STATE_DELTAS]: {
    numPartitions: 1,
    replicationFactor: 1,
    retentionMs: 7 * 24 * 60 * 60 * 1000,
  },
  [TOPICS.METRICS_OPERATIONAL]: {
    numPartitions: 1,
    replicationFactor: 1,
    retentionMs: 30 * 24 * 60 * 60 * 1000, // 30 days for metrics
  },
};

File: domains/ingestion/realtime/src/topics.rs
//! Redpanda topic name constants — the canonical topic names established by
//! ADR 0008 (Adopt Redpanda as Immutable Temporal Snapshot Ledger). See
//! `docs/design/redpanda-topic-configuration.md` for the full topology.
//!
//! Only `SNAPSHOTS_RAW` is populated today; the remaining topics are defined
//! here for completeness and so string literals don't get scattered across
//! the codebase.

pub const SNAPSHOTS_RAW: &str = "transit.snapshots.raw";
pub const SNAPSHOTS_NORMALIZED: &str = "transit.snapshots.normalized";
pub const STATE_DELTAS: &str = "transit.state.deltas";
pub const METRICS_OPERATIONAL: &str = "transit.metrics.operational";

#[derive(Debug, Clone, Copy)]
pub struct TopicConfig {
    pub name: &'static str,
    pub num_partitions: i32,
    pub replication_factor: i16,
    /// Retention, in milliseconds.
    pub retention_ms: i64,
}

const DAY_MS: i64 = 24 * 60 * 60 * 1000;

/// Phase 1 uses 1 partition per topic intentionally — ADR 0008 explicitly
/// defers partitioning specialisation until replay patterns are validated
/// empirically.
pub const TOPIC_CONFIGS: &[TopicConfig] = &[
    TopicConfig {
        name: SNAPSHOTS_RAW,
        num_partitions: 1,
        replication_factor: 1,
        retention_ms: 7 * DAY_MS,
    },
    TopicConfig {
        name: SNAPSHOTS_NORMALIZED,
        num_partitions: 1,
        replication_factor: 1,
        retention_ms: 7 * DAY_MS,
    },
    TopicConfig {
        name: STATE_DELTAS,
        num_partitions: 1,
        replication_factor: 1,
        retention_ms: 7 * DAY_MS,
    },
    TopicConfig {
        name: METRICS_OPERATIONAL,
        num_partitions: 1,
        replication_factor: 1,
        // Longer retention for operational/analytics workloads.
        retention_ms: 30 * DAY_MS,
    },
];

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn snapshots_raw_retains_seven_days() {
        let cfg = TOPIC_CONFIGS
            .iter()
            .find(|c| c.name == SNAPSHOTS_RAW)
            .unwrap();
        assert_eq!(cfg.retention_ms, 604_800_000);
        assert_eq!(cfg.num_partitions, 1);
        assert_eq!(cfg.replication_factor, 1);
    }

    #[test]
    fn metrics_operational_retains_thirty_days() {
        let cfg = TOPIC_CONFIGS
            .iter()
            .find(|c| c.name == METRICS_OPERATIONAL)
            .unwrap();
        assert_eq!(cfg.retention_ms, 2_592_000_000);
    }

    #[test]
    fn every_phase_1_topic_has_a_config() {
        let names: Vec<&str> = TOPIC_CONFIGS.iter().map(|c| c.name).collect();
        assert_eq!(
            names,
            vec![
                SNAPSHOTS_RAW,
                SNAPSHOTS_NORMALIZED,
                STATE_DELTAS,
                METRICS_OPERATIONAL
            ]
        );
    }
}
